python/ray/tune/examples/mnist_tune_serve.py

```diff
@@ -0,0 +1,498 @@
+    }
+
+    scheduler = ASHAScheduler(
+        metric="mean_accuracy",
+        mode="max",
+        max_t=num_epochs,
+        grace_period=1,
+        reduction_factor=2)
+
+    reporter = CLIReporter(
+        parameter_columns=["layer_size", "lr", "momentum", "batch_size"],
+        metric_columns=["mean_accuracy", "training_iteration"])
+
+    analysis = tune.run(
+        partial(
+            train_from_scratch,
+            data_interface=data_interface,
+            num_epochs=num_epochs,
+            day=day),
+        resources_per_trial={
+            "cpu": 1,
+            "gpu": gpus_per_trial
+        },
+        config=config,
+        num_samples=num_samples,
+        scheduler=scheduler,
+        progress_reporter=reporter,
+        checkpoint_at_end=True,
+        verbose=0,
+        name="tune_serve_mnist_fromscratch")
+
+    best_trial = analysis.get_best_trial("mean_accuracy", "max", "last")
+    best_accuracy = best_trial.metric_analysis["mean_accuracy"]["last"]
+    best_trial_config = best_trial.config
+    best_checkpoint = best_trial.checkpoint.value
+
+    return best_accuracy, best_trial_config, best_checkpoint, N
+
+
+def tune_from_existing(start_model,
+                       start_config,
+                       num_samples=10,
+                       num_epochs=10,
+                       gpus_per_trial=0.,
+                       day=0):
+    data_interface = MNISTDataInterface("/tmp/mnist_data", max_days=10)
+    N = data_interface._get_day_slice(day) - data_interface._get_day_slice(
+        day - 1)
+
+    config = start_config.copy()
+    config.update({
+        "batch_size": tune.choice([16, 32, 64]),
+        "lr": tune.loguniform(1e-4, 1e-1),
+        "momentum": tune.uniform(0.1, 0.9),
+    })
+
+    scheduler = ASHAScheduler(
+        metric="mean_accuracy",
+        mode="max",
+        max_t=num_epochs,
+        grace_period=1,
+        reduction_factor=2)
+
+    reporter = CLIReporter(
+        parameter_columns=["lr", "momentum", "batch_size"],
+        metric_columns=["mean_accuracy", "training_iteration"])
+
+    analysis = tune.run(
+        partial(
+            train_from_existing,
+            start_model=start_model,
+            data_interface=data_interface,
+            num_epochs=num_epochs,
+            day=day),
+        resources_per_trial={
+            "cpu": 1,
+            "gpu": gpus_per_trial
+        },
+        config=config,
+        num_samples=num_samples,
+        scheduler=scheduler,
+        progress_reporter=reporter,
+        checkpoint_at_end=True,
+        verbose=0,
+        name="tune_serve_mnist_fromsexisting")
+
+    best_trial = analysis.get_best_trial("mean_accuracy", "max", "last")
+    best_accuracy = best_trial.metric_analysis["mean_accuracy"]["last"]
+    best_trial_config = best_trial.config
+    best_checkpoint = best_trial.checkpoint.value
+
+    return best_accuracy, best_trial_config, best_checkpoint, N
+
+
+if __name__ == "__main__":
+    """
```

Comment:
let's put this at top of file?

Reply:
I would actually like to keep it at the end. For one, it is quite conventional to have the main entrypoint at the end of a python script. But the main reason I see is that this code looks a bit cluttered and off-putting compared to the rest of the code, and calls functions would not yet be explained. I think it's much nicer to have everything layed otu/explained before and then have the glue code at the end.

Reply:
Oh I meant the docstring to go to the head of the file

Reply:
Ah, got it. I think I'll put some of the usage examples from it to the top so people now what the goal for the script is